mkt/extensions/models.py

```diff
@@ -140,18 +155,52 @@ def is_public(self):
     def manifest_url(self):
         raise NotImplementedError
 
-    @property
-    def path_prefix(self):
-        return os.path.join(settings.ADDONS_PATH, 'extensions', str(self.pk))
-
     def publish(self):
         """Publish this add-on to public."""
-        # FIXME: sign and move file to public storage.
+        self.sign_and_move_file()
         self.update(status=STATUS_PUBLIC)
 
     def reject(self):
         """Reject this add-on."""
         self.update(status=STATUS_REJECTED)
+        self.remove_signed_file()
+
+    def remove_signed_file(self):
+        if public_storage.exists(self.signed_file_path):
+            public_storage.delete(self.signed_file_path)
+
+    def sign_and_move_file(self):
+        """Sign and move extension file from the unsigned path (`file_path`) on
+        private storage to the signed path (`signed_file_path`) on public
+        storage."""
+        if not self.uuid:
+            raise SigningError('Need uuid to be set to sign')
+
+        ids = json.dumps({
+            # 'id' needs to be an unique identifier not shared with anything
+            # else (other extensions, langpacks, webapps...), but should not
+            # change when there is an update.
+            'id': self.uuid,
+            # 'version' should be an integer and should be monotonically
+            # increasing.
+            'version': self.file_version
+        })
+        with statsd.timer('extensions.sign'):
+            try:
+                # This will read the file from self.file_path, generate a
+                # signature and write the signed file to self.signed_file_path.
+                sign_app(private_storage.open(self.file_path),
+                         self.signed_file_path, ids)
```

Comment:
Yeah, it doesn't care about the manifest, and they are both zips, signed the same way. It *should* work on the device, but I really have no idea at this point : I need to have the manifest and signing working to test it.